sklearn/preprocessing/data.py

```diff
@@ -2209,6 +2209,17 @@ def __init__(self, n_quantiles=1000, output_distribution='uniform',
         self.random_state = random_state
         self.copy = copy
 
+    @staticmethod
+    def _nanpercentile_force_finite(a, q):
+        """Force the output of nanpercentile to be finite."""
+        with warnings.catch_warnings():
+            warnings.filterwarnings('ignore')
+            percentile = nanpercentile(a, q)
+            if np.all(np.isclose(percentile, np.nan, equal_nan=True)):
+                return np.zeros(len(q))
```

Comment:
We should respect `a.dtype` that is `return np.zeros(len(q), dtype=a.dtype)` to avoid upcasting `np.float32` input features.